app/controllers/courses_controller.rb

```diff
@@ -43,27 +43,35 @@ def switch_role
     if params[:effective_user_login].blank?
       render partial: 'role_switch_handler',
              formats: [:js], handlers: [:erb],
-             locals: { error: I18n.t('main.username_not_blank') }
+             locals: { error: I18n.t('main.username_not_blank') },
+             status: :not_found
       return
     end
 
     found_user = User.find_by_user_name(params[:effective_user_login])
     found_role = Role.find_by(end_user: found_user, course: current_course)
 
-    if found_role.nil? || found_role.instructor?
+    if found_role.nil?
       render partial: 'role_switch_handler',
              formats: [:js], handlers: [:erb],
-             locals: { error: Settings.validate_user_not_allowed_message || I18n.t('main.login_failed') }
+             locals: { error: Settings.validate_user_not_allowed_message || I18n.t('main.login_failed') },
+             status: :not_found
       return
     end
-
-    # Check if an instructor trying to login as the current user or themselves
+    # check if the current instructor is trying to role switch as themselves
     if found_user.user_name == session[:user_name] || found_user.user_name == session[:real_user_name]
```

Comment:
Because the `role_switch` policy checks that the user is currently not role-switched, the check against `session[:user_name]` can be removed.